tests/foreman/api/test_discoveryrule.py

```diff
@@ -23,221 +24,91 @@
-        self.assertEqual(loc.name, discovery_rule.location[0].read().name)
-
-    @tier1
-    def test_positive_delete(self):
-        """Delete a discovery rule
-
-        :id: 9fdba953-dcc7-4532-9204-17a45b0d9e05
-
-        :expectedresults: Rule should be deleted successfully
-
-        :CaseImportance: Critical
-        """
-        for name in valid_data_list():
-            with self.subTest(name):
-                self.discovery_rule.name = name
-                discovery_rule = self.discovery_rule.create()
-                discovery_rule.delete()
-                with self.assertRaises(HTTPError):
-                    discovery_rule.read()
-
-    @tier1
-    def test_negative_create_with_too_long_name(self):
-        """Create a discovery rule with more than 255 char in name
-
-        :id: 415379b7-0134-40b9-adb1-2fe0adb1ac36
-
-        :expectedresults: Validation error should be raised
-        """
-        for name in (
-            gen_string(str_type, 256) for str_type in ('alpha', 'numeric', 'alphanumeric')
-        ):
-            with self.subTest(name):
-                self.discovery_rule.name = name
-                with self.assertRaises(HTTPError):
-                    self.discovery_rule.create()
-
-    @tier1
-    def test_negative_create_with_invalid_host_limit(self):
-        """Create a discovery rule with invalid host limit
-
-        :id: 84503d8d-86f6-49bf-ab97-eff418d3e3d0
-
-        :expectedresults: Validation error should be raised
-        """
-        self.discovery_rule.max_count = gen_string('alpha')
-        with self.assertRaises(HTTPError):
-            self.discovery_rule.create()
-
-    @tier1
-    def test_negative_create_with_invalid_priority(self):
-        """Create a discovery rule with invalid priority
-
-        :id: 4ec7d76a-22ba-4c3e-952c-667a6f0a5728
-
-        :expectedresults: Validation error should be raised
-        """
-        self.discovery_rule.priority = gen_string('alpha')
-        with self.assertRaises(HTTPError):
-            self.discovery_rule.create()
-
-    @tier1
-    def test_positive_update_name(self):
-        """Update an existing discovery rule name
-
-        :id: 769c0739-538b-4451-af7b-deb2ecd3dc0d
-
-        :expectedresults: User should be able to update the rule
-
-        :CaseImportance: Critical
-        """
-        discovery_rule = self.discovery_rule.create()
-        for name in valid_data_list():
-            with self.subTest(name):
-                discovery_rule.name = name
-                discovery_rule = discovery_rule.update(['name'])
-                self.assertEqual(discovery_rule.name, name)
-
-    @tier1
-    def test_positive_update_org_loc(self):
-        """Update org and location of selected discovery rule
-
-        :id: 0f8ec302-f9de-4713-87b7-0f1aca515149
-
-        :expectedresults: Rule was updated and with given org & location
-        """
+        assert name == discovery_rule.name
```

Comment:
I don't see any teardown of the created entities in this test pattern.  The test eventually calls delete on the discovery rule - but that only happens if all of the interim assertions pass.

I think after creating these entities, we should use `request.addfinalizer(entity.delete)` so that on teardown (success or failure within the case) the entities are cleaned up after the test is done.

If the creation of these entities (other than discovery rule, as that is the target of the test) is moved into a fixture, then you don't have to use `addfinalizer` and you can clean them up after yielding.

Reply:
wouldn't `request.addfinalizer(entity.delete)` crash the whole pytest session in case the server fails to delete it? Or will it be treated same as if the fixture failed? just asking.

I prefer the cleanup to be put into a fixture.

Reply:
iirc there already were attempts to tear down the taxonomies after test execution, but due the way Satellite used to treat them*, the teardown failed too often.
It would definitely be worth revisiting this topic to explore what has changed since then.
Such inconsistencies in behaviour were reported and a lot of them got fixed eventually.
A lot of seemingly weird behaviour was also caused by using "admin" user account, which is not limited to selected subset of taxonomies. perhaps having a (non admin) session user for every test (created vy a fixture) might deliver the needed isolation as well as more accurate UX. This is a broad topic deserving its own issue ticket and should be handled at design level .
We probably don't want to block this PR.

*e.g. - a test1 created org1 for its own use. In the meantime, test2 created a Smart Proxy entity, which automatically got assigned to all existing orgs. After that, test1 tried to tear down the org1 which satellite refused to do as "it was used by the Smart Proxy".

Reply:
if you're happy with this and consider the comment resolved, please Merge the PR. it already got all the required Approvals.
thanks!

Reply:
I do agree with @rplevka this is with the whole project and needs discussion and design implementation steps. We can plan a similar #7686 strategy to every where